### DISCUSSION TITLE
Force tailwind config to reload on external module plugin change

### DISCUSSION THREAD

**Discussion by greven** [tailwindlabs/tailwindcss#10213] [Help] [answered]
Hey! Thanks for any help in advance.

I'm starting to write a UI Kit a la DaisyUI where there is an external module function that is exported where addComponents, etc is called.

This file is then consumed by a tailwind.config.js file as a plugin, nothing out of the ordinary.

```js
const plugin = require('tailwindcss/plugin')
const components = require('../dist/components')

const plugins = ({ addComponents }) => {
  addComponents(components)
}
    
module.exports = plugin(plugins, {})
```

Then on the tailwind config file:

```js
    module.exports = {
      content: ['./src/**/*.{astro,html,svelte,vue,js,ts,jsx,tsx}'],
      plugins: [require('../src')],
    }
```

The limitation I found with this setup is that if I change the external component code (a components.json file with the CSSinJS code) the tailwind config doesn't reload itself, which makes sense as I don't think tailwind keeps tracks of external / "dependency" / required files. 

If I inline the components into the plugin section of the tailwind config file it all just reloads as expected (as used in the examples in the Tailwind docs).

My question is, is anything I can do to force the reload of the tailwind config file so in development I can just change the CSS and see the changes reflected without having to restart the server that calls the tailwind config file?
I guess this could be solved if there was a hook to trigger a config reload or something.

Thanks!

**Comment by benweier** [comment]
Hi @greven, I've been in the same situation for the last couple months and decided to see what I could do about it. I have just published the first iteration of a module I'm calling [`tailwind-apply`](https://www.npmjs.com/package/tailwind-apply) that does exactly what I need it to do. It's rather simple - just two files - and if you find this doesn't do what you need it to do, I welcome feedback and suggestions or you're free to copy and adapt it for your own requirements.

As you have figured out, using `addComponents` inside `tailwind.config.js` works just fine, but using an external plugin file and making changes in it doesn't reload the styles. You either need to restart the build or manually force a change to `tailwind.config.js`. Turns out, it technically isn't _directly_ tied to TailwindCSS but PostCSS instead.

The `tailwindcss` plugin informs PostCSS that `tailwind.config.js` is considered a dependency and when changes are detected PostCSS handles reloading the file. Unfortunately any imports required inside `tailwind.config.js` are not included in the PostCSS dependencies list. So how do we do that?

The `tailwind-apply/postcss` plugin is designed to read all matching files in a folder and add them to the dependencies list to trigger a CSS build on change [benweier/tailwind-apply/postcss](https://github.com/benweier/tailwind-apply/blob/main/postcss.js). This has to be done before the `tailwindcss` plugin. The following example matches all `css` and `js` files in your `src/styles` folder.

```js
module.exports = {
  plugins: [
    require('tailwind-apply/postcss')({ styles: './src/styles', pattern: '**/*.{css,js}' }),
    require('tailwindcss'),
    require('autoprefixer'),
  ],
}
```

The second plugin for TailwindCSS is only necessary _if you author your custom classes with CSS/SCSS/Sass/Less_, but when authoring in a JS file the value passed to `addComponents` is likely to be a JS object already (matching CSS files are converted to a JS object by this plugin using `postcss-js`). I recommend that simple custom classes use JS files but more complex classes - especially those that make use of `@apply`! - should definitely use CSS files.

Requiring this file inside the `tailwindcss` plugins array works so you only need `tailwind-apply/postcss` to watch for file changes.

```js
const plugin = require('tailwindcss/plugin')

module.exports = plugin(({ addComponents }) => {
  addComponents({
    '.btn': {
      padding: '.5rem 1rem',
      borderRadius: '.25rem',
      fontWeight: '600',
    },
  })
})
```

Since this css file cannot be required in `tailwindcss` plugins, use the `tailwind-apply` plugin to have it parsed into JS and `tailwind-apply/postcss` to watch for change.

```css
.btn {
  @apply rounded bg-gradient-to-r from-pink-500 via-fuchsia-500 to-blue-500;
  @apply py-2.5 px-4 text-sm font-semibold text-white;
  @apply hover:from-pink-600 hover:via-fuchsia-600 hover:to-blue-600;
  @apply focus:outline-none focus:ring-2 focus:ring-zinc-700 focus:ring-offset-2 focus:ring-offset-zinc-100;
  @apply disabled:cursor-not-allowed disabled:opacity-70;
}
```

I hope this helps you in some way!

**Comment by benweier** [comment, marked as answer]
Shoot forgot to include documentation about the `styles` folder structure.

If using CSS files the styles should be arranged into folders named after the layer they target

```
project/
├── src
│   ├── components / etc   <-- your projects standard folders
│   └── index.tsx
└── styles   <-- or wherever you save your custom styles
    ├── base   <-- styles that get added to Tailwind's `base` layer
    ├── components   <-- styles that get added to Tailwind's `component` layer
    │   ├── button.css
    │   └── input.css
    ├── utilities   <-- added to `utilities` layer
    │   └── content.css
    └── variants   <-- you get the idea
```

**Comment by greven** [comment]
Hey @benweier That's great! I started investigating if it was Tailwind or PostCSS but since I'm overwhelmed with worked didn't do much more than start to look into it.
This looks great, I'm going to try it out on my small personal project where I found this limitation to see if it works.

Thanks a lot for your work on this!
